### PR TITLE
Add support for injecting a banner in HTML responses

### PR DESCRIPTION
**Pull request by ojarjur**
This can be useful for adding functionality to a web-app without having to change the backend server.

### CODE REVIEW

File: agent/banner/banner.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+			return
+		}
+		if isAlreadyFramed(r) {
+			// The page is already framed
+			wrapped.ServeHTTP(w, r)
+			return
+		}
+		responseRecorder := httptest.NewRecorder()
+		wrapped.ServeHTTP(responseRecorder, r)
+		result := responseRecorder.Result()
```
Comment (bsidhom) [inline agent/banner/banner.go]: It seems a bit weird to use httptest for this purpose. Could we somehow inspect/transform the results in-line? How do the other proxy layers do this (or is the answer that they effectively return all responses unchanged)?

Comment (ojarjur) [reply agent/banner/banner.go]: Fair enough. I've changed it to modify the response in-line.

File: agent/agent.go [hunk 2]
```diff
@@ -70,10 +72,18 @@ func hostProxy(ctx context.Context, host, shimPath string, injectShimCode bool)
 		Host:   host,
```
[the remark this answers is not in the document]
Comment (ojarjur) [reply agent/agent.go]: That's a good point.

Currently, every user of this I know of is just connecting to a server on localhost, so they are always http, but there is no reason to require that.

File a tracking issue for that [here](https://github.com/google/inverting-proxy/issues/37)

File: agent/banner/banner.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	accept := r.Header.Get(acceptHeader)
+	return strings.Contains(accept, "html")
+}
+
+func isHTMLResponse(resp *http.Response) bool {
+	if resp.StatusCode != http.StatusOK {
+		return false
+	}
+	contentType := resp.Header.Get(contentTypeHeader)
+	return strings.Contains(contentType, "html")
```
[the remark this answers is not in the document]
Comment (ojarjur) [reply agent/banner/banner.go]: Done.

File: agent/banner/banner.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    {{.Banner}}
+    <iframe height="100%" width="100%" style="border:0px" src="{{.TargetURL}}"></iframe>
+  </body>
+</html>`
+)
+
+var frameWrapperTmpl = template.Must(template.New("frame-wrapper").Parse(frameWrapperTemplate))
+
+func isHTMLRequest(r *http.Request) bool {
+	if r.Method != http.MethodGet {
```
[the remark this answers is not in the document]
Comment (ojarjur) [reply agent/banner/banner.go]: I thought through this a bit and came to the conclusion that we should not inject the banner into responses for POSTS.

My concern is that an app might send a POST request and not be resilient to the response being modified.

As such, I think we should err on the side of not injecting the banner.

I've added comments to the code explaining this.

**Pull request merged**